lib/compress/zstd_compress.c

```diff
@@ -203,6 +207,38 @@ size_t ZSTD_sizeof_CStream(const ZSTD_CStream* zcs)
 /* private API call, for dictBuilder only */
 const seqStore_t* ZSTD_getSeqStore(const ZSTD_CCtx* ctx) { return &(ctx->seqStore); }
 
+/* Returns if the strategy uses the row based matchfinder */
+static int ZSTD_rowMatchFinderSupported(const ZSTD_strategy strategy) {
+    return (strategy >= ZSTD_greedy && strategy <= ZSTD_lazy2);
+}
+
+/* Returns row matchfinder usage enum given an initial mode, cParams, srcSize */
+static ZSTD_useRowMatchFinderMode_e ZSTD_resolveRowMatchFinderMode(ZSTD_useRowMatchFinderMode_e mode,
+                                                                   const ZSTD_compressionParameters* const cParams) {
+#if !defined(ZSTD_NO_INTRINSICS) && (defined(__SSE2__) || defined(__ARM_NEON))
+    int const kHasSIMD128 = 1;
+#else
+    int const kHasSIMD128 = 0;
+#endif
+    if (mode != ZSTD_urm_auto) return mode;
+    mode = ZSTD_urm_disableRowMatchFinder;
+    if (!ZSTD_rowMatchFinderSupported(cParams->strategy)) return mode;
+    if (kHasSIMD128) {
+        if (cParams->windowLog > 14) mode = ZSTD_urm_enableRowMatchFinder;
+    } else {
+        if (cParams->windowLog > 17) mode = ZSTD_urm_enableRowMatchFinder;
+    }
+    return mode;
+}
+
+/* Returns 1 if the arguments indicate that we should allocate a chainTable, 0 otherwise */
+static int ZSTD_allocateChainTable(const ZSTD_strategy strategy, const ZSTD_useRowMatchFinderMode_e useRowMatchFinder, const U32 forCCtx) {
+    assert(useRowMatchFinder != ZSTD_urm_auto);
+    return (strategy != ZSTD_fast) && (useRowMatchFinder == ZSTD_urm_disableRowMatchFinder
+                                    || !ZSTD_rowMatchFinderSupported(strategy)
+                                    || !forCCtx);
```

Comment:
Why is this `!forCCtx`?

Reply:
Hmm, this does seem unnecessary: I initially had thought that re-use of a static cdict between DDS and DMS might get screwy if the chainTable isn't allocated if we switch from DDS->DMS (and some seemingly relevant `fuzzer.c` tests were failing), but it seems like this should be okay now that DMS is fixed.